infra/Pos/Diffusion/Subscription/Dns.hs

```diff
@@ -107,10 +115,24 @@ dnsSubscriptionWorker oq networkCfg DnsDomains{..} keepaliveTimer nextSlotDurati
 
     -- How long to wait before retrying in case no alternative can be
     -- subscribed to.
-    retryInterval :: m Millisecond
-    retryInterval = do
+    --
+    -- The formula to calculate the retry interval is
+    -- @
+    -- retryInterval = floor (slotDuration / 2 ^ (number of slots that passed during the longest subscription duration))
+    -- @
+    -- For a @20s@ slot: the values will be @20s@, @10s@, @5s@, @2s@,
+    -- @1s@, @0s@, i.e. if there was a subscription that lasted more than
+    -- @5@ slots we will try to re-subscribe immediately.
+    retryInterval :: Millisecond -> m Millisecond
+    retryInterval d = do
         slotDur <- nextSlotDuration
-        pure $ max (slotDur `div` 4) (convertUnit (5 :: Second))
+        let -- slot duration in microseconds
+            slotDurF :: Float
+            slotDurF = fromIntegral $ toMicroseconds slotDur
+            -- Number of slots that passed in the longest subscription duration
+            slotsF :: Float
+            slotsF = (fromIntegral $ toMicroseconds d) / slotDurF
+        return $ fromMicroseconds $ floor $ slotDurF / (2 ** slotsF)
```

Comment:
From meeting today: @karknu pointed out that `getPOSIXTime` is not necessarily monotonic. It's possible that the system clock changes mid-subscription and gives a negative `d`, leading to a negative `slotsF`, which could give a massive value here. Either we use a monotonic clock, or just throw in a `max 1` here.

Reply:
The `System.Clock.Monotonic` module provides platform independent monotonic clock with this two issues (which in my opinion will not hurt our software):
 
>  * On Windows XP, this uses GetTickCount, which has a 49.7 day wraparound. Clock works around this problem, but the workaround only works if clockGetTime is called at least once every 24.8 days.
> * On Linux, this uses clock_gettime with CLOCK_MONOTONIC, which (unfortunately) stops when the computer is suspended. Thus, clockGetTime will not include time spent sleeping. Do not rely on this behavior, as it may be fixed in a future version of this library.

Reply:
Good to know. I can only imagine the problems we might have if the node is kept running through a suspend... would be interesting to try it out.

Reply:
The same happens with `System.Clock.getTime Monotonic` from the `clock` library, at least on linux. It stops running when I suspend my laptop.